src/widgets/scrollbar.rs

```diff
@@ -0,0 +1,606 @@
+pub struct ScrollbarState {
+    // The current position within the scrollable content.
+    position: u16,
+    // The total length of the scrollable content.
+    content_length: u16,
+}
+
+impl ScrollbarState {
+    /// Sets the scroll position of the scrollbar and returns the modified ScrollbarState.
+    pub fn position(mut self, position: u16) -> Self {
+        self.position = position;
+        self
+    }
+
+    /// Sets the length of the scrollable content and returns the modified ScrollbarState.
+    pub fn content_length(mut self, content_length: u16) -> Self {
+        self.content_length = content_length;
+        self
+    }
+
+    /// Decrements the scroll position by one, ensuring it doesn't go below zero.
+    pub fn prev(&mut self) {
+        self.position = self.position.saturating_sub(1);
+    }
+
+    /// Increments the scroll position by one, ensuring it doesn't exceed the length of the content.
+    pub fn next(&mut self) {
+        self.position = self
+            .position
+            .saturating_add(1)
+            .clamp(0, self.content_length.saturating_sub(1))
+    }
+
+    /// Sets the scroll position to the start of the scrollable content.
+    pub fn first(&mut self) {
+        self.position = 0;
+    }
+
+    /// Sets the scroll position to the end of the scrollable content.
+    pub fn last(&mut self) {
+        self.position = self.content_length.saturating_sub(1)
+    }
+
+    /// Changes the scroll position based on the provided ScrollDirection.
+    pub fn scroll(&mut self, direction: ScrollDirection) {
+        match direction {
+            ScrollDirection::Forward => {
+                self.next();
+            }
+            ScrollDirection::Backward => {
+                self.prev();
+            }
+        }
+    }
+}
+
+/// Scrollbar Orientation
+#[derive(Default, Debug, Clone)]
+pub enum ScrollbarOrientation {
+    #[default]
+    Vertical,
+    Horizontal,
+}
+
+/// Scrollbar widget for tui-rs library.
+///
+/// This widget can be used to display a scrollbar in a terminal user interface.
+#[derive(Debug, Clone)]
+pub struct Scrollbar<'a> {
+    pub orientation: ScrollbarOrientation,
+    pub thumb_style: Style,
+    pub thumb_symbol: &'a str,
+    pub track_style: Style,
+    pub track_symbol: &'a str,
+    pub begin_symbol: Option<&'a str>,
+    pub begin_style: Style,
+    pub end_symbol: Option<&'a str>,
+    pub end_style: Style,
+}
+
+impl<'a> Default for Scrollbar<'a> {
+    fn default() -> Self {
+        Self {
+            orientation: ScrollbarOrientation::default(),
+            thumb_symbol: DOUBLE_VERTICAL.thumb,
+            thumb_style: Style::default(),
+            track_symbol: DOUBLE_VERTICAL.track,
+            track_style: Style::default(),
+            begin_symbol: None, // Some(DOUBLE_VERTICAL.begin),
+            begin_style: Style::default(),
+            end_symbol: None, // Some(DOUBLE_VERTICAL.end),
+            end_style: Style::default(),
```

Comment:
Should arrows be enabled by default?

Reply:
I'm leaning toward enabling arrows by default. 

One reason is when using a scrollbar with the same size of the rect, the arrows take up the corners:

<img width="733" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/d051c1b8-c6a8-4f6c-b9ea-a05a784db763">

when is a better out of the box experience than this:

<img width="732" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/f020bc25-8753-4913-9fe1-f11166fd30ec">

Without arrows, most users will want to write this:

```
        Scrollbar::default()
            .orientation(ScrollbarOrientation::Vertical)
            .arrows(false),
        chunks[0].inner(&Margin {
            vertical: 1,
            horizontal: 0,
        })
```

i.e. decrease the size of the scrollbar by 1 so that it looks nicer:

<img width="726" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/ac130c09-1f80-40d3-be6d-812853762fc0">

That is a less nice first experience for a user.

Reply:
+1 to arrows by default.

I'd be inclined to say that for the scrollbar to render as part of a paragraph that has borders, it should be inside the borders not on the borders (but also that this should be configurable - I like the idea of being able to use the border as a scroll bar like what you have here).

Reply:
The scrollbar widget here doesn’t know anything about other widgets. 

Are you imaging that another widget like a  `Paragraph` would store an instance to a scrollbar and manipulate where it is being rendered?

Reply:
Btw, I like the scrollbar in the border because it reuses space that isn’t useful (inside it may overwrap on text or eat into the width or height of a text area), and because it can be turned off and on without any reflow on content in the widget.

Reply:
yup - I'm thinking about how existing widgets will use the scrollbar.
The would create one in their render() method.

Reply:
That makes sense! Happy to add any other functions you think will improve the interface for other widgets or users.